std/algorithm/searching.d

```diff
@@ -1773,44 +1773,85 @@ if (isRandomAccessRange!R1 && hasLength!R1 && hasSlicing!R1 && isBidirectionalRa
         && is(typeof(binaryFun!pred(haystack.front, needle.front)) : bool))
 {
     if (needle.empty) return haystack;
-    const needleLength = walkLength(needle.save);
-    if (needleLength > haystack.length)
+    static if (hasLength!R2)
     {
-        return haystack[haystack.length .. haystack.length];
+        immutable needleLength = needle.length;
     }
-    // @@@BUG@@@
-    // auto needleBack = moveBack(needle);
-    // Stage 1: find the step
-    size_t step = 1;
-    auto needleBack = needle.back;
-    needle.popBack();
-    for (auto i = needle.save; !i.empty && i.back != needleBack;
-         i.popBack(), ++step)
+    else
     {
+        immutable needleLength = walkLength(needle.save);
```

Comment:
But you're already testing for hasLength; only ranges with lengths will be sent to walkLength.